### PR TITLE
 Implement subclass compatibility for workflow pipeline job steps

### PR DESCRIPTION
**Pull request by jerrypeng7773**
*Issue #, if available:*

*Description of changes:*

 Implement subclass compatibility for workflow pipeline job steps. Please reach out to the Pipeline team before merge, there are few other PRs we would like to release together.

*Testing done:*

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I certify that the changes I am introducing will be backword compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [x] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [x] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/sagemaker/workflow/steps.py [hunk 1]
```diff
@@ -524,6 +570,10 @@ def __init__(
         super(ProcessingStep, self).__init__(
             name, StepTypeEnum.PROCESSING, display_name, description, depends_on, retry_policies
         )
+        if not (run_args is None or processor is None):
+            raise ValueError("either run_args or processor need to be given.")
+
```
Comment (qidewenwhen) [inline src/sagemaker/workflow/steps.py:575]: The case is missing that both `run_args` and `processor` are None.
May need to check other steps as well.

Comment (jerrypeng7773) [reply src/sagemaker/workflow/steps.py:577]: yep, fixed.

File: src/sagemaker/processing.py [hunk 2]
```diff
@@ -314,10 +313,10 @@ def _normalize_inputs(self, inputs=None, kms_key=None):
                 if file_input.input_name is None:
                     file_input.input_name = "input-{}".format(count)
 
-                if isinstance(file_input.source, Properties) or file_input.dataset_definition:
+                if isinstance(file_input.source, PipelineVariable) or file_input.dataset_definition:
```
[the remark this answers is not in the document]
Comment (jerrypeng7773) [reply src/sagemaker/processing.py:316]: sure

File: src/sagemaker/workflow/steps.py [hunk 3]
```diff
@@ -270,22 +271,40 @@ def __init__(
         super(TrainingStep, self).__init__(
             name, StepTypeEnum.TRAINING, display_name, description, depends_on, retry_policies
         )
+
+        if (run_args and estimator) or not (run_args is None or estimator is None):
+            raise ValueError("either run_args or estimator need to be given.")
+
```
[the remark this answers is not in the document]
Comment (jerrypeng7773) [reply src/sagemaker/workflow/steps.py:277]: you are right, fixing.

File: src/sagemaker/workflow/steps.py [hunk 4]
```diff
@@ -76,7 +76,6 @@ class Step(Entity):
         step_type (StepTypeEnum): The type of the `Step`.
         depends_on (List[str] or List[Step]): The list of `Step` names or `Step`
             instances that the current `Step` depends on.
-        retry_policies (List[RetryPolicy]): The custom retry policy configuration.
```
[the remark this answers is not in the document]
Comment (jerrypeng7773) [reply src/sagemaker/workflow/steps.py:79]: coz this is a property belongs to `ConfigurableRetryStep`

File: src/sagemaker/workflow/pipeline_context.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+            run_func(*args, **kwargs)
+            return args[0].sagemaker_session.context
+
+        run_func(*args, **kwargs)
+        return None
+
+    return wrapper
+
+
+def is_pipeline_entities(obj: Any) -> bool:
```
[the remark this answers is not in the document]
Comment (jerrypeng7773) [reply src/sagemaker/workflow/pipeline_context.py:136]: yep.

File: src/sagemaker/workflow/steps.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jerrypeng7773) [reply src/sagemaker/workflow/steps.py:277]: so, this is XOR, we can do if not (run_args ^ estimator) here.

**Review by qidewenwhen — APPROVED** [review]

File: src/sagemaker/session.py [hunk 6]
```diff
@@ -4161,6 +4175,15 @@ def account_id(self) -> str:
         )
         return sts_client.get_caller_identity()["Account"]
 
+    def _intercept_create_request(self, request: typing.Dict, create):
+        """This function intercepts the create job request
```
Comment (mufaddal-rohawala) [inline src/sagemaker/session.py:4179]: Whats the change we are trying to accomplish in this module?

**Comment by mufaddal-rohawala** [comment]
@[jerrypeng7773](https://github.com/jerrypeng7773) is this PR is not ready to merge, lets convert this PR to draft.

File: src/sagemaker/session.py (hunk 6 above)
Comment (jerrypeng7773) [reply src/sagemaker/session.py:4179]: Unlike normal SM session, under pipeline session, we need to capture the create job request, and stop. This function will be overrode in `PipelineSession` class to achieve this.

**Comment by jerrypeng7773** [comment]
> @[jerrypeng7773](https://github.com/jerrypeng7773) is this PR is not ready to merge, lets convert this PR to draft.

this PR is ready to merge, but let me ping another team member to give it another scan.

File: src/sagemaker/workflow/pipeline_context.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    def context(self):
+        """Hold contextual information useful to the session"""
+        return self._context
+
+    @context.setter
+    def context(self, args: Dict):
+        self._context = args
+
+    def _intercept_create_request(self, request: Dict, create):
+        self.context = request
```
Comment (qidewenwhen) [inline src/sagemaker/workflow/pipeline_context.py:77]: What is the `create` used for? Don't see any usage of it in this method.

Comment (qidewenwhen) [reply src/sagemaker/workflow/pipeline_context.py:77]: NVM, found it's just an override method

File: src/sagemaker/session.py (hunk 6 above)
Comment (jerrypeng7773) [reply src/sagemaker/session.py:4179]: sure

**Review by navaj0 — APPROVED** [review]
This is a copy of another PR, which I have commented. All comments are addressed.

**Review by qidewenwhen — APPROVED** [review]

**Review by navaj0 — APPROVED** [review]

**Review by ahsan-z-khan — APPROVED** [review]

**Review by mufaddal-rohawala — APPROVED** [review]

**Pull request merged**